src/hotspot/os_cpu/linux_riscv/threadLS_linux_riscv.S

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2015, Red Hat Inc. All rights reserved.
+// Copyright (c) 2021, Alibaba Group Holding Limited. All Rights Reserved.
+// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+//
+// This code is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License version 2 only, as
+// published by the Free Software Foundation.
+//
+// This code is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// version 2 for more details (a copy is included in the LICENSE file that
+// accompanied this code).
+//
+// You should have received a copy of the GNU General Public License version
+// 2 along with this work; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+// or visit www.oracle.com if you need additional information or have any
+// questions.
+
+        // JavaThread::riscv64_get_thread_helper()
+        //
+        // Return the current thread pointer in a0(x10).
+        // All other registers are preserved,
+
+    .global _ZN10JavaThread25riscv64_get_thread_helperEv
+    .type   _ZN10JavaThread25riscv64_get_thread_helperEv, %function
+
+_ZN10JavaThread25riscv64_get_thread_helperEv:
```

Comment:
For maximum flexibility, I would suggest we switch to use the global-dynamic TLS model here if you don't have a strong reason to use the initial-exec TLS model. That's the default TLS model used by the host GCC compiler and I don't want to affected by the limits of the initial-exec TLS model some day.
---------------
        la.tls.gd       a0, _ZN6Thread12_thr_currentE
        addi    sp, sp, -16
        sd      ra, 8(sp)
        call    __tls_get_addr@plt
        ld      ra, 8(sp)
        ld      a0, 0(a0)
        addi    sp, sp, 16
        ret
---------------

Reply:
Hi Felix -

Quite sorry for the late reply - it takes me some time to look up some references. This code looks great without any problem and I totally agree with your point - so it is hard to make me not think you have implemented one version already in your codebase :-)  And `__tls_get_addr`'s implementation could found itself at [this link](https://sources.debian.org/src/glibc/2.31-13/sysdeps/riscv/libc-tls.c/) with [macros defined](https://sources.debian.org/src/glibc/2.31-13/sysdeps/riscv/nptl/tls.h/), which is a C function. With this respect, it seems that I also need to save and restore all volatile registers pessimistically in `MacroAssembler::get_thread`, same as the current implementation on the master branch. I think it might be hard to keep the saved registers the same as AArch64's implementation that only clobbers x1 - please correct me if I missed something. But this seems much cheaper than calling a [pthread_getspecific](https://sources.debian.org/src/glibc/2.32-4/nptl/pthread_getspecific.c/?hl=24#L24). In fact, the first version of this patch is written in a similar way as yours - for better performance I chose to use tls.ie - but I think you hold a very reasonable point.

It's Friday night so please feel free to move the discussion to weekdays.

Thanks -
Xiaolin

Reply:
Seems `get_thread()` is only used in debug code on AArch64 and RISCV -- I have pushed revised code and now hotspot and jdk tier1 are under testing. I will ping this thread when tests are done.

Reply:
Updated changes looks good. Let's see when you have the test result.